test/hotspot/jtreg/compiler/loopopts/TestUnswitchPredicateCloning.java

```diff
@@ -0,0 +1,185 @@
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package compiler.loopopts;
+
+import compiler.lib.ir_framework.*;
+import java.util.Random;
+import jdk.test.lib.Utils;
+import jdk.test.lib.Asserts;
+
+/*
+ * @test
+ * @bug 8346552
+ * @summary Test that all parse predicates are cloned after loop unswitching.
+ * @key randomness
+ * @library /test/lib /
+ * @run driver compiler.loopopts.TestUnswitchPredicateCloning
+ */
+
+public class TestUnswitchPredicateCloning {
+    static final int SIZE = 100;
+
+    private static final Random random = Utils.getRandomInstance();
+
+    public static void main(String[] strArr) {
+        TestFramework.run();
+    }
+
+    @Run(test = {"testUnswitchingBeforePredication", "testPredicationBeforeUnswitching", "testUnswitchingUncounted"})
+    @Warmup(0)
+    private static void runNoWarmup() {
+        final int idx = random.nextInt(SIZE);
+        final boolean cond = random.nextBoolean();
+        int res = testUnswitchingBeforePredication(idx);
+        Asserts.assertEQ(SIZE * idx, res);
+        res = testPredicationBeforeUnswitching(idx, cond);
+        Asserts.assertEQ((SIZE * (SIZE - 1)) / 2 + (cond ? SIZE * idx : 0), res);
+        res = testUnswitchingUncounted(cond);
+        Asserts.assertEQ((SIZE * (SIZE - 1)) / 2 + (cond ? SIZE : 0), res);
+    }
+
+    @DontInline
+    private static int[] getArr() {
+        int[] arr = new int[SIZE];
+        for (int i = 0; i < SIZE; i++) {
+            arr[i] = i;
+        }
+
+        return arr;
+    }
+
+    @Test
+    // Check that Loop Unswitching doubled the number of Parse Predicates: We have
+    // them at the true- and false-path-loop. Note that the Loop Limit Check Parse
+    // Predicate is not cloned when we already have a counted loop.
+    @IR(counts = { IRNode.LOOP_PARSE_PREDICATE, "3",
+                   IRNode.PROFILED_LOOP_PARSE_PREDICATE, "3",
+                   IRNode.LOOP_LIMIT_CHECK_PARSE_PREDICATE, "3",
+                   IRNode.AUTO_VECTORIZATION_CHECK_PARSE_PREDICATE, "3" },
+        phase = CompilePhase.BEFORE_LOOP_UNSWITCHING)
+    // Since we know that predication happens after unswitching, we can test the
+    // predicate cloning before predication, such that the useless, killed predicates
+    // have already been removed in the beautify loop phase.
+    @IR(counts = { IRNode.LOOP_PARSE_PREDICATE, "4",
+                   IRNode.PROFILED_LOOP_PARSE_PREDICATE, "4",
+                   IRNode.LOOP_LIMIT_CHECK_PARSE_PREDICATE, "3",
+                   IRNode.AUTO_VECTORIZATION_CHECK_PARSE_PREDICATE, "4" },
+        phase = CompilePhase.BEFORE_LOOP_PREDICATION_IC)
```

Comment:
I suggest to match on RC to have matching before/after RC pairs
```suggestion
        phase = CompilePhase.BEFORE_LOOP_PREDICATION_RC)
```